api: Create PUT /stream/pull API for idempotent pull stream

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.studio/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

This creates a separate `PUT /stream/pull` API to replace the `autoStartPull` query-string
that had been created on the regular creation API initially.

This is because the semantics expected from this pull API are now becoming very different
from the regular POST-based Create Stream API that we have today. Specifically, the 2 new
requirements are:
- It should update the stream object if it already exists
- It should block the request until the stream actually starts on the media server

Since blocking a connection is a non-standard and non-recommended behavior, that part
was still kept under a query-string. So this API has 2 possible idempotence results:
- If `waitActive` is NOT specified: when the API succeeds, the stream is guaranteed to be created and the pull to have been triggered.
- If `waitActive` IS specified: when the API succeeds, the stream is guaranteed to be created, the pull triggered, and the media server already started the stream.

In both cases, if a stream with the same `pull.source` already existed, it will have been updated
and then triggered to start.

One caveat is that we make no checks about the stream _already_ being active when this API is called.
In that case, the media server won't actually update the configuration of the ongoing stream, so it is an
exception to the idempotence aimed by this API. We can think of alternatives, like returning an error when
the stream is already active, similar to the PATCH API, but it depends on what the design partner for this
feature expects from this API.

Also notice that I ended up preferring not to use this: https://github.com/livepeer/livepeer-data/pull/180
Doing a simpler poll on the DB instead for the `stream.isActive` field.

**Specific updates (required)**
- Abstract the "create stream" logic from the existing API into a helper, to be reusable in the new API
- Create separate `PUT /stream/pull` API
- Implement logic to update stream object based on PUT payload
- Implement logic to wait for stream to become active

**How did you test each of these updates (required)**
The pull logic on Catalyst is not implement yet so it's not really possible to test it. Mostly tested calling the API itself.

**Does this pull request close any open issues?**
Implements latest requirements here https://discord.com/channels/423160867534929930/1201860223652859986/1201860228413399040

**Checklist**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

victorges committed:
api/schema: Make lat/lon required fields

victorges committed:
api: Create separate PUT /stream/pull endpoint

victorges committed:
api: Filter by userId and non-deleted on queries

scary

victorges committed:
api: Update stream fields on PUT /pull API

victorges committed:
api: Wait until stream is active after trigger

victorges committed:
api: Return 200 when the stream already existed

victorges committed:
api: Allow querying by creatorId/pull.source

victorges committed:
api: Allow deduping streams by creatorId

victorges committed:
api/test: Fix asset indexes test

victorges committed:
api/test: Add tests for new API

victorges committed:
api: Fix response not to omit default fields

We had an object with a couple of undefined fields
which then didn't get the default added by addDefaultFields

Fix it by re-reading the obj from the DB, which will not only
get the actually serialized version (undefined is omited) but
get the actual final state in the db (maybe someone wrote at
the same time :shrug:)

thomshutt committed:
Merge branch 'master' into vg/feat/blocking-stream-pull

mjh1 reviewed (approved):
Are you thinking you'll remove the stream pull logic from the create stream handler in a later PR?

victorges committed:
api: Fix indexes

Remove 'name' from /pull potential keys. It doesn't
have an index and we won't need it anyway for now, so
let's KISS.

Also moved index: directive to db-schema
